modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/actionItems/filterItemActions.ts

```diff
@@ -3,58 +3,82 @@
  */
 
-import {IItemsActions} from '../..';
+import {IItemActionsDataFilter, IItemsActions} from '../..';
+import {getLocalizedValue} from '../getLocalizedValue';
+
+const hasPermission = (action: IItemsActions, itemData: any): boolean => {
+	if (!action?.data?.permissionKey) {
+		return true;
+	}
+
+	if (!itemData?.actions) {
+		return false;
+	}
+
+	const permissionKey = action.data?.permissionKey?.toLowerCase();
+
+	return Object.keys(itemData?.actions).some(
+		(itemAction) => itemAction.toLowerCase() === permissionKey
+	);
+};
+
+const matchesFilters = (action: IItemsActions, itemData: any): boolean => {
+	if (!action?.data?.filters) {
+		return true;
+	}
+
+	return action?.data?.filters.every(
+		(filter: IItemActionsDataFilter) =>
+			getLocalizedValue(itemData, filter?.key)?.value === filter?.value
+	);
+};
+
+const isVisible = (action: IItemsActions, itemData: any): boolean => {
+	if (!hasPermission(action, itemData) || !matchesFilters(action, itemData)) {
+		return false;
+	}
+
+	if (action?.isVisible) {
+		return action.isVisible(itemData);
+	}
+
+	return true;
+};
+
+const transformAction = (
+	action: IItemsActions,
+	itemData: any
+): IItemsActions => {
+	if (!action?.data?.permissionKey || action?.target !== 'headless') {
+		return action;
+	}
+
+	const permissionKey = action?.data?.permissionKey?.toLowerCase();
+
+	const matchedPermissionKeys = Object.keys(itemData?.actions).filter(
+		(itemAction) => itemAction.toLowerCase() === permissionKey
+	);
+
+	if (!matchedPermissionKeys.length) {
+		return action;
+	}
+
+	return {
+		...action,
+		...itemData?.actions[matchedPermissionKeys[0]],
+	};
+};
 
 const filterItemActions = (
 	actions: Array<IItemsActions>,
 	itemData: any
 ): Array<IItemsActions> => {
 	return actions
-		? actions.reduce(
-				(actions: Array<IItemsActions>, action: IItemsActions) => {
-					if (action.data?.permissionKey) {
-						const itemDataActionKeys = Object.keys(
-							itemData.actions
-						);
-
-						if (
-							itemData.actions &&
-							itemDataActionKeys.some(
-								(itemAction) =>
-									itemAction.toLowerCase() ===
-									action.data?.permissionKey?.toLowerCase()
-							)
-						) {
-							if (action.target === 'headless') {
-								const matchedPermissionKey =
-									itemDataActionKeys.filter(
-										(itemAction) =>
-											itemAction.toLowerCase() ===
-											action.data?.permissionKey?.toLowerCase()
-									);
-
-								return [
-									...actions,
-									{
-										...action,
-										...itemData.actions[
-											matchedPermissionKey[0]
-										],
-									},
-								];
-							}
-							else {
-								return [...actions, action];
-							}
-						}
-
-						return actions;
-					}
-
-					return [...actions, action];
-				},
-				[]
-			)
+		? actions
+				.filter((action: IItemsActions) => isVisible(action, itemData))
+				.map((action: IItemsActions) =>
+					transformAction(action, itemData)
+				)
```

Comment:
I love the split! :+1:  I always found the big reduce to be too dense and hard to understand.